File: baekjoon/7주차/1629-곱셈/우섭/1629.py
from sys import stdin
# a ^ b % m
def pow(a, b, m):
    if b == 1:
        return a % m
    val = pow(a, b//2, m)
    val = val * val % m
    if (b % 2 == 0):
        return val
    return val * a % m

a, b, m = map(int, stdin.readline().split(' '))
print(pow(a, b, m))

File: baekjoon/7주차/1629-곱셈/제문/1629.kt
package baekjoon

import java.lang.StringBuilder

private val out = StringBuilder()
private val br = System.`in`.bufferedReader()

private fun readLine() = br.readLine()!!
private fun readInt() = br.readLine().toInt()
private fun readInts() = br.readLine().split(" ").map { it.toInt() }
private fun readLongs() = br.readLine().split(" ").map { it.toLong() }

fun main() {
    val (A, B, C) = readLongs()

    print(multiple(A, B, C))
}

fun multiple(A: Long, B: Long, C: Long): Long {
    if (B == 1L) return A % C
    var result = multiple(A, B/2, C)
    result = result * result % C

    return if (B % 2 == 0L) result else result * A % C
}